client/my-sites/plugins/plugins-list/index.jsx

```diff
@@ -0,0 +1,402 @@
+		this.setState( { selectedPlugins: Object.assign( {}, selectedPlugins, { [ slug ]: ! oldValue } ) } );
+		analytics.ga.recordEvent( 'Plugins', 'Clicked to ' + this.isSelected( plugin ) ? 'Deselect' : 'Select' + 'Single Plugin', 'Plugin Name', slug );
+	},
+
+	setBulkSelectionState( plugins, selectionState ) {
+		let slugsToBeUpdated = {};
+		plugins.forEach( plugin => slugsToBeUpdated[ plugin.slug] = this.hasNoSitesThatCanManage( plugin ) ? false : selectionState );
+
+		this.setState( { selectedPlugins: Object.assign( {}, this.state.selectedPlugins, slugsToBeUpdated ) } );
+	},
+
+	getPluginBySlug( slug ) {
+		for ( let i = 0, l = this.props.plugins.lenght; i < l; i++ ) {
+			if ( this.props.plugins[ i ].slug === slug ) {
+				return this.props.plugins[ i ];
+			}
+		}
+		return false;
+	},
+
+	filterSelection: {
+		active( plugin ) {
+			if ( this.isSelected( plugin ) ) {
+				return plugin.sites.some( site => site.plugin && site.plugin.active );
+			}
+			return false;
+		},
+		inactive( plugin ) {
+			if ( this.isSelected( plugin ) ) {
+				return plugin.sites.some( site => site.plugin && ! site.plugin.active );
+			}
+			return false;
+		},
+		updates( plugin ) {
+			if ( this.isSelected( plugin ) ) {
+				return plugin.sites.some( site => site.plugin && site.plugin.update && site.canUpdateFiles );
+			}
+			return false;
+		},
+		selected( plugin ) {
+			return this.isSelected( plugin );
+		}
+	},
+
+	getPluginUpdateCount() {
+		if ( this.props.isWpCom ) {
+			return 0;
+		}
+		return this.props.pluginUpdateCount;
+	},
+
+	hasNoSitesThatCanManage( plugin ) {
+		return ! plugin.wpcom &&
+			! plugin.sites.some( site => site.modules && site.modules.indexOf( 'manage' ) !== -1 );
+	},
+
+	getSelected() {
+		if ( ! this.props.plugins ) {
+			return [];
+		}
+		return this.props.plugins.filter( this.filterSelection.selected.bind( this ) );
+	},
+
+	siteSuffix() {
+		return ( this.props.selectedSite || this.props.sites.get().length === 1 ) ? '/' + this.props.sites.selected : '';
+	},
+
+	recordEvent( eventAction, includeSelectedPlugins ) {
+		eventAction += ( this.props.selectedSite ? '' : ' on Multisite' );
+		if ( includeSelectedPlugins ) {
+			const pluginSlugs = this.getSelected().map( plugin => plugin.slug );
+			analytics.ga.recordEvent( 'Plugins', eventAction, 'Plugins', pluginSlugs );
+		} else {
+			analytics.ga.recordEvent( 'Plugins', eventAction );
+		}
+	},
+
+	// Actions
+	toggleBulkManagement() {
+		const bulkManagement = ! this.state.bulkManagement;
+
+		this.setState( {
+			bulkManagement
+		} );
+
+		if ( bulkManagement ) {
+			this.recordEvent( 'Clicked Manage' );
+		} else {
+			// Unselect all plugins.
+			this.setState( { selectedPlugins: {} } );
+
+			//PluginsActions.selectPlugins( this.props.sites, 'none' );
+			if ( this.state.notices && ( this.state.notices.completed || this.state.notices.errors ) ) {
+				PluginsActions.removePluginsNotices( this.state.notices.completed.concat( this.state.notices.errors ) );
```

Comment:
this is another case where the _soft failure_ mode of destructuring clears things up quite a bit

```js
const { notices: { completed, errors } = {} } = this.state;

if ( completed || errors ) {
	…
}
```